### PR TITLE
Improve compat with pre-dataclass expressions

### PR DESCRIPTION
**Pull request by alexfikl**
Mostly adds back some `dofdesc=None` default constructors found during merging main into #162.

### CODE REVIEW

File: pytential/symbolic/mappers.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ def rec_int_g_arguments(mapper, expr):
             name: mapper.rec(arg) for name, arg in expr.kernel_arguments.items()
             }
 
-    changed = (
+    changed = not (
```
Comment (alexfikl) [inline pytential/symbolic/mappers.py:64]: I do not know how this didn't break the world before.. :facepalm:

File: pytential/symbolic/mappers.py [hunk 2]
```diff
@@ -280,7 +280,12 @@ def map_common_subexpression(self, expr):
 # {{{ FlattenMapper
 
 class FlattenMapper(FlattenMapperBase, IdentityMapper):
-    pass
+    def map_int_g(self, expr):
+        densities, kernel_arguments, changed = rec_int_g_arguments(self, expr)
+        if not changed:
+            return expr
+
+        return replace(expr, densities=densities, kernel_arguments=kernel_arguments)
```
Comment (alexfikl) [inline pytential/symbolic/mappers.py:288]: Also from #162, it would be nice to teach `ExpressionKernel` in sumpy to flatten a bit more. Not sure what the best place for that is..

Comment (inducer) [inline pytential/symbolic/mappers.py:288]: Why is this needed? It looks an awful lot like what would be inherited from `IdentityMapper`? (If there is a reason, please add a comment.)

Comment (alexfikl) [reply pytential/symbolic/mappers.py]: Yeah, it is. It didn't work due to https://github.com/inducer/pytential/pull/248#discussion_r1847138517 and I forgot to revert this part.

Removed in https://github.com/inducer/pytential/compare/165d0e140a2538ededcbcf1545b4c21205f27e34..baba46fa9af2222a69b1816438274a443e1baa32

File: pytential/symbolic/mappers.py (hunk 1 above)
Comment (inducer) [reply pytential/symbolic/mappers.py:64]: Agreed. Scary... Thanks for finding it!
